File: PiecewiseConstant/synthesize/plots/probability.m
% Plotting ibp-interval bounds on erf functions

clc; clear; close;

% Probability distribution on hyperspace
partitions = load('../models/linear/state_partitions.txt');
data_hyper = load('../models/linear/probability_data_5_sigma_0.01.mat');
prob_lower = data_hyper.matrix_prob_lower;
prob_upper = data_hyper.matrix_prob_upper;

figure
hold on
grid on

% Note: this example is for transition from Xj to Xi

for jj = 1:length(partitions)
    
    colors = {'b', 'k', 'm', 'r', 'c'};
    for ii = 3%1:1%length(partitions)
    
        x_space = linspace(min(partitions(jj, :)), max(partitions(jj, :)), 1000);
        
        % True erf
        prob_true = zeros(1, length(x_space));
        sigma = 0.01;
        epsilon = 1e-6;
        m = 1;      % sys dim
        const = 1/(2^m);

        vl = min(partitions(ii, :));
        vu = max(partitions(ii, :));

        % Prod of erfs
        for pp = 1:length(x_space)
            y = 0.95*x_space(pp);
            erf_low = (y - vl)/(sigma*sqrt(2));
            erf_up = (y - vu)/(sigma*sqrt(2));
            prob_true(pp) = const*(erf(erf_low) - erf(erf_up));
            % Notice this computes Pu --> Pu = 1 - Ps
        end

        min_prob_exact = prob_lower(jj, ii);
        max_prob_exact = prob_upper(jj, ii);

        plot(x_space, prob_true, "LineWidth", 3, 'Color', "r")

        plot(x_space, min_prob_exact*ones(1, length(x_space)), "LineWidth", 3, 'Color', "b")
        plot(x_space, max_prob_exact*ones(1, length(x_space)), "LineWidth", 3, 'Color', "k")
 
    end
end

labels = {'\color{red} true', '\color{blue} lower bound', ...
          '\color{black} upper bound'};
legend(labels, 'Location', 'NorthEast', 'FontSize', 8, ...
'TextColor', 'black'); 
xlabel("$x$", 'Interpreter','latex', "FontSize", 25)
ylabel('$P$', 'Interpreter','latex', "FontSize", 25)
title("Probability distribution")
set(gcf,'color','w');
set(gca,'FontSize',14)

File: PiecewiseConstant/synthesize/plots/probability.jl
using Revise
using Plots, DelimitedFiles, MAT
using SpecialFunctions: erf

# Plotting ibp-interval bounds on erf functions

# Probability distribution on hyperspace
partitions = readdlm(joinpath(@__DIR__, "../models/linear/state_partitions.txt"))
data_hyper = matopen(joinpath(@__DIR__, "../models/linear/probability_data_5_sigma_0.01.mat"))
prob_lower = read(data_hyper, "matrix_prob_lower")
prob_upper = read(data_hyper, "matrix_prob_upper")
prob_unsafe_lower = read(data_hyper, "matrix_prob_unsafe_lower")
prob_unsafe_upper = read(data_hyper, "matrix_prob_unsafe_upper")
close(data_hyper)


# Note: this example is for transition from Xⱼ to Xᵢ

f(x) = 1.05 * x
sigma = 0.01
m = 1      # sys dim

for ii in axes(partitions, 1)
    colors = [:blue, :black, :magenta, :red, :cyan]

    x = range(minimum(partitions[:, 1]), maximum(partitions[:, 2]), length = 1000)
        
    # True prob
    vl = partitions[ii, 1]
    vu = partitions[ii, 2]

    y = f.(x)

    erf_lower = @. (y - vl) / (sigma * sqrt(2))
    erf_upper = @. (y - vu) / (sigma * sqrt(2))
    prob_true = (1 / 2^m) .* (erf.(erf_lower) - erf.(erf_upper))

    p = plot(x, prob_true, linewidth = 3, color = :red, label="true")
    
    for jj in axes(partitions, 1)
        x_space = partitions[jj, :]
        
        plot!(p, x_space, fill(prob_lower[ii, jj], 2), linewidth = 3, color = :blue, label=jj == 1 ? "lower bound" : nothing)
        plot!(p, x_space, fill(prob_upper[ii, jj], 2), linewidth = 3, color = :black, label=jj == 1 ? "upper bound" : nothing)
    end

    savefig(p, "transition_prob$ii.png")
end

# xlabel!("x")
# ylabel!("P")
# title!("Probability distribution")
